src/mame/konami/konamigv.cpp

```diff
@@ -772,36 +832,201 @@ uint16_t tokimeki_state::tokimeki_serial_r()
+	if (printer_bit && !m_printer_bit)
+	{
+		m_printer_pulse_starttime = machine().time();
+	}
+	else if (!printer_bit && m_printer_bit)
+	{
+		/*
+		duration of the transition from 1->0->1 is used to figure out what is being sent/
+		the durations are clocked against the PS1's clock via GetRCnt(0xf2000002)
+		internally the values compared are 2400, 4800, and 9600 cycles
+		9600 = start of new data transfer
+		4800 = data bit 1
+		2400 = data bit 0
+		*/
+		const auto curtime = machine().time();
+		const auto ticks = (curtime - m_printer_pulse_starttime).as_ticks(m_maincpu->clock() / 16);
+
+		if (ticks >= 9000)
+		{
+			// start of data transfer
+			m_printer_curbit = 0;
+		}
+		else
+		{
+			// data bit transfer
+			const int bit = ticks >= 4000 && ticks <= 5600;
+
+			if (m_printer_curbit == 0)
+				m_printer_data[m_printer_curbyte] = 0;
+
+			m_printer_data[m_printer_curbyte] |= bit << m_printer_curbit;
+			m_printer_curbit++;
+
+			// game is programmed to always sends two bytes. first byte is 7 bits, second byte is hardcoded to be 0xf9
+			if ((m_printer_curbyte & 1) == 0 && m_printer_curbit == 7)
+			{
+				m_printer_curbyte++;
+				m_printer_curbit = 0;
+			}
+			else if ((m_printer_curbyte & 1) == 1 && m_printer_curbit == 8)
+			{
+				if (m_printer_data[m_printer_curbyte] != 0xf9)
+					logerror("tokimeki printer 2nd byte was expected to be f9, found %02x", m_printer_data[m_printer_curbyte]);
+
+				m_printer_curbyte++;
+				m_printer_curbit = 0;
+			}
+		}
+
+		if (m_printer_curbyte >= 6)
+		{
+			/*
+			All of the commands seem to correspond to physical buttons available on the machine.
+			The manual configuration option in the printer test menu appears to let you control
+			the printer's OSD via the cabinet controls and you can see it sending 64/65/66/67
+			depending on which direction you press on the joystick.
+
+			0x0b = sent before 0x63 -> 0x62 sequence when holding button to exit manual configuration menu. stop?
+			0x10 = memory in
+			0x11 = print
+			0x17 = source/memory?
+			0x62 = menu
+			0x63 = exec
+			0x64 = up
+			0x65 = down
+			0x66 = left
+			0x67 = right
+			*/
+
+			// The same sequence of 2 bytes will be sent 3 times every time it is sent, so only accept the value if it was repeated 3 times
+			if (m_printer_data[0] != m_printer_data[2] || m_printer_data[0] != m_printer_data[4] || m_printer_data[1] != m_printer_data[3] || m_printer_data[1] != m_printer_data[5])
+			{
+				logerror("printer command not accepted, found different bytes [%02x %02x %02x] [%02x %02x %02x]", m_printer_data[0], m_printer_data[2], m_printer_data[4], m_printer_data[1], m_printer_data[3], m_printer_data[5]);
+			}
+			else if (m_printer_data[1] != 0xf9)
+			{
+				logerror("printer command not accepted, found second byte that wasn't 0xf9 [%02x %02x %02x] [%02x %02x %02x]", m_printer_data[0], m_printer_data[2], m_printer_data[4], m_printer_data[1], m_printer_data[3], m_printer_data[5]);
+			}
+			else if (m_printer_data[0] == 0x10)
+			{
```

Comment:
My gut feel is that this should wait for vertical blanking to ensure it gets a properly rendered frame.  Can you start a timer on receiving this message to wait until some arbitrary point early in the vertical blanking interval to ensure you grab a frame after the final screen update for the frame happens?

In real life, the printer is going to have to see a complete frame after receiving this command (i.e. wait for current frame to complete, then start capturing at the top of the next frame).

Reply:
Does this seem sane for what you're suggesting? https://github.com/mamedev/mame/pull/12233/commits/eab9298c6f0e8a835085ab0997cbdf683bfe2f98

Reply:
Yeah, looks about right.